# Diabetes Prediction Deep Learning Model

# Problem Statement : 

The objective of the Pima Indian diabetes dataset is to diagnostically predict whether a patient has diabetes based on certain diagnostic information included in the dataset. All the patients are female and of Pima Indian heritage. 

In [1]:
# Import the libraries you need

from keras.models import Sequential
from keras.layers import Dense
import numpy

C:\Users\PSMike\Anaconda3\lib\site-packages\h5py\__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
Using TensorFlow backend.


In [2]:
# Import dataset to view headers

import pandas as pd
dataset_csv = pd.read_csv('pimac.csv') 
dataset_csv.head()

TimesPregnat  Glucose  BloodPressure  SkinThickness  Insulin   BMI    DPG  \
0             6      148             72             35        0  33.6  0.627   
1             1       85             66             29        0  26.6  0.351   
2             8      183             64              0        0  23.3  0.672   
3             1       89             66             23       94  28.1  0.167   
4             0      137             40             35      168  43.1  2.288   

   Age  TargetVariable  
0   50               1  
1   31               0  
2   32               1  
3   21               0  
4   33               1

In [3]:
# fix random seed for reproducibility

numpy.random.seed(7)

# load pima indians dataset

dataset = numpy.loadtxt("pima.txt", delimiter=",")

In [4]:
dataset

array([[  6.   , 148.   ,  72.   , ...,   0.627,  50.   ,   1.   ],
       [  1.   ,  85.   ,  66.   , ...,   0.351,  31.   ,   0.   ],
       [  8.   , 183.   ,  64.   , ...,   0.672,  32.   ,   1.   ],
       ...,
       [  5.   , 121.   ,  72.   , ...,   0.245,  30.   ,   0.   ],
       [  1.   , 126.   ,  60.   , ...,   0.349,  47.   ,   1.   ],
       [  1.   ,  93.   ,  70.   , ...,   0.315,  23.   ,   0.   ]])

In [5]:
# split into input (X) and output (Y) variables

X = dataset[:,0:8]
Y = dataset[:,8]

In [6]:
# create model

model = Sequential() # define keras model
model.add(Dense(12, input_dim=8, activation='relu')) # adds 1st layer of model- the input layer; 12 represents the number of neurons in your layer; input_dim is the number of attributes you are passing from your dataset; uses Relu activation function.
model.add(Dense(8, activation='relu')) 
model.add(Dense(1, activation='sigmoid')) # output layer needs only 1 neuron because our model is a binary classification problem so the output is 1 or 0

In [7]:
# Compile model

model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy']) # passing the loss function, an optimizer and a metric 'accuracy' for your model

# Fit the model

model.fit(X, Y, epochs=150, batch_size=10) # 1 epoch is when the entire dataset is passed through the neural network once; batch_size is the total no. of training examples present in single batch


Epoch 1/150
768/768 [==============================] - 0s 400us/step - loss: 3.6841 - acc: 0.5977
Epoch 2/150
768/768 [==============================] - 0s 85us/step - loss: 0.9278 - acc: 0.6003
Epoch 3/150
768/768 [==============================] - 0s 87us/step - loss: 0.7406 - acc: 0.6354
Epoch 4/150
768/768 [==============================] - 0s 93us/step - loss: 0.7083 - acc: 0.6602
Epoch 5/150
768/768 [==============================] - 0s 84us/step - loss: 0.6810 - acc: 0.6849
Epoch 6/150
768/768 [==============================] - 0s 83us/step - loss: 0.6501 - acc: 0.6849
Epoch 7/150
768/768 [==============================] - 0s 79us/step - loss: 0.6485 - acc: 0.6732
Epoch 8/150
768/768 [==============================] - 0s 84us/step - loss: 0.6357 - acc: 0.6927
Epoch 9/150
768/768 [==============================] - 0s 82us/step - loss: 0.6232 - acc: 0.6914
Epoch 10/150
768/768 [==============================] - 0s 79us/step - loss: 0.6279 - acc: 0.6797
Epoch 11/150
768/768 [======

In [8]:
# evaluate the model

scores = model.evaluate(X, Y)
print("\n%s: %.2f%%" % (model.metrics_names[1], scores[1]*100)) # print the accuracy of the model

768/768 [==============================] - 0s 55us/step

acc: 78.39%


# Conclusion:

 The accuracy of our deep learning binary classification model is 78.39% which is pretty good actually. 